### DISCUSSION TITLE
train_grpo.py中torch+grpo路径下old_per_token_logps未detach导致GRPO优势项梯度抵消

### DISCUSSION THREAD

**Discussion by zeta172417-design** [jingyaogong/minimind#788] [General]
## 问题描述

我在使用`train_grpo.py`训练时，发现当使用如下配置：

```bash
--rollout_engine torch
--loss_type grpo
--use_moe 0
```

也就是`torch+dense+grpo`路径时，`old_per_token_logps`和`per_token_logps`可能来自同一个未detach的张量，导致GRPO中的ratio虽然数值为1，但梯度被抵消，进而使reward/advantage项无法有效更新actor。

---

## 相关代码

当前代码中大致是这样：

```python
old_per_token_logps = rollout_result.per_token_logps.to(args.device)

...

if use_sglang or lm_config.use_moe:
    res = model_unwrapped(outputs)
    aux_loss = res.aux_loss if lm_config.use_moe else torch.tensor(0.0, device=args.device)
    logits = res.logits[:, :-1, :]
    per_token_logps = F.log_softmax(logits, dim=-1).gather(
        2, outputs[:, 1:].unsqueeze(-1)
    ).squeeze(-1)[:, -completion_ids.size(1):]
else:
    aux_loss = torch.tensor(0.0, device=args.device)
    per_token_logps = rollout_result.per_token_logps
```

在`torch+dense+grpo`路径下，会走`else`分支，于是：

```python
old_per_token_logps = rollout_result.per_token_logps
per_token_logps = rollout_result.per_token_logps
```

然后GRPO中计算：

```python
ratio = torch.exp(per_token_logps - old_per_token_logps)
```

等价于：

```python
ratio = torch.exp(x - x)
```

这时`ratio`数值上等于1，但因为是同一个计算图张量相减，所以梯度也会被抵消：

```text
d(x - x) / dx = 0
```

因此GRPO的reward/advantage项可能无法有效更新actor，训练主要只剩下KL项在起作用。

---

## 建议修改

可以将`old_per_token_logps`显式detach：

```python
old_per_token_logps = rollout_result.per_token_logps.detach().clone().to(args.device)
```

这样ratio变成：

```python
ratio = torch.exp(x - x.detach())
```

数值上初始仍然是1，但梯度不会被抵消：

```text
d(x - x.detach()) / dx = 1
```

这样GRPO的advantage项才能正常通过当前策略的`per_token_logps`更新actor。

---

## 为什么这个问题主要出现在torch+dense+grpo路径

我理解是：

1. `sglang`路径下，`old_per_token_logps`来自外部推理服务，本身不连接本地PyTorch计算图；
2. `use_moe`路径下，会重新forward计算当前`per_token_logps`，不再是同一个张量直接相减；
3. 但`torch+dense`路径下，当前`per_token_logps`直接复用`rollout_result.per_token_logps`，如果`old_per_token_logps`不detach，就会出现`exp(x - x)`导致梯度抵消的问题。

---

## 观察到的现象

修改前，`loss_type=grpo`时模型似乎较难从reward/advantage中学习，`kl_ref`变化也较小。

将`old_per_token_logps`改为：

```python
old_per_token_logps = rollout_result.per_token_logps.detach().clone().to(args.device)
```

之后，`kl_ref`变化明显增大，说明actor相对ref model开始发生更明显的策略偏移，这也符合GRPO优势项梯度恢复后的预期。

---

## 想确认的问题

请问这里`old_per_token_logps`未detach是有意设计，还是一个可能的实现bug？

从GRPO/PPO的理论上看，`old_per_token_logps`应该表示采样时旧策略的固定logprob，因此似乎应当作为常数处理，不应该继续参与反向传播。

**Comment by jingyaogong** [comment]
代码早已改为“一律重新 forward + 显式 .detach()”，该 Bug 不复存在，你没有把fork或者本地的代码对齐版本
